### ISSUE TITLE
Question about encoder output shape when decoding with beam search

### ISSUE THREAD

**Issue by jellysnack** [NVIDIA/FasterTransformer#405]
Judging by the decoding code, the first dimension of the encoder output is expected to be batch * beam. I tried running three different options:
1) Duplicate the encoder output beam times
2) Do not duplicate the encoder output, but allocated additional memory according to batch * beam size
3) Do not duplicate the encoder output and do not allocate additional memory, but only change the dimension in the input tensor to batch * beam

All three options worked successfully in debug mode and gave the correct result. However, when run with cuda compute-sanitizer, the third option failed with some errors.

Tell me please, which option is correct in order to start decoding with beam search?

**Comment by byshiue** [comment]
Can you point out which example you use? The memory allocation are hidden in FT. Where do you allocate addition memory? 

From design view, the output of encoder should only have bs, and we will duplicate it in decoding.

**Comment by jellysnack** [comment]
I'm talking about code from `src/fastertransformer/models/decoding`
Both this code and `examples/cpp/decoding/decoding_example.cc` indicate that the first dimension of the encoder output must be equal to `batch * beam`.  I assumed it was a typo, because it is strange that the shape of the encoder output depend on the beam size. However, if the beam size is removed from the input tensors in the `examples/cpp/decoding/decoding_example.cc` example, then a core dump occurs in the cross attention layer.

Here is a code example that reproduces the error (this is a slightly modified `decoding_example.cc`): https://gist.github.com/jellysnack/2ae9c038dd06606bdcc9453fcfbccf66
I ran the example with this command: `./decoding_example 16 2 16 64 2048 16000 6 32 80 1024 1 0 1`

debug log
[log.txt](https://github.com/NVIDIA/FasterTransformer/files/10309550/log.txt)

GPU: A100
Driver Version: 470.57.02
CUDA Version: 11.7

**Comment by jellysnack** [comment]
I was talking about the code `src/fastertransformer/models/decoding` and the corresponding example `examples/cpp/decoding/decoding_example.cc`

I was confused by the fact that the first dimension of the encoder output must be equal to `batch * beam` and experimented with different options, assuming that this could be a typo. Now I looked at the corresponding python example and ensured that I actually need to copy encoder output according to beam size. And in the cpp code of the T5 model I found that I need to use the `invokeTileEncoderResults` function.

It seems that the `invokeTileEncoderResults` function should also be added to the decoding code. If it helps you, I can make a pull request.
